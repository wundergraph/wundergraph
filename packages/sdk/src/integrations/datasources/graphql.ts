import type { GraphQLIntrospection } from '../../definition';
import { defineDatasource } from './define-datasource';

export interface GraphQLDatasourceOptions extends Omit<GraphQLIntrospection, 'apiNamespace'> {
	namespace?: string;
}

export const graphql = defineDatasource<GraphQLDatasourceOptions>((config) => {
	const { namespace, ...introspectionConfig } = config;
	return {
		name: 'graphql',
		hooks: {
			'config:setup': async (options) => {
				const { introspect } = await import('../../definition');
				options.addApi(
					introspect.graphql({
						apiNamespace: namespace,
						...introspectionConfig,
					})
				);
			},
		},
	};
});
