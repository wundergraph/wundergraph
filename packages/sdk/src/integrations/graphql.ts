import type { GraphQLIntrospection } from '../definition';
import { defineDatasource } from './define-datasource';

export interface GraphQLDatasourceOptions extends Omit<GraphQLIntrospection, 'apiNamespace'> {
	namespace: string;
}

export const graphql = defineDatasource<GraphQLDatasourceOptions>((config) => {
	return {
		name: 'graphql',
		hooks: {
			'config:setup': async (options) => {
				const { introspect } = await import('../definition');
				options.addApi(
					introspect.graphql({
						apiNamespace: config.namespace,
						...config,
					})
				);
			},
		},
	};
});
