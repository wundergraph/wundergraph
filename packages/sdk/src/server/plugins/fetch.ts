import { FastifyBaseLogger } from 'fastify';

type FetchFn = (input: RequestInfo | URL, init?: RequestInit) => Promise<Response>;

/**
 * Create a fetch function that logs its requests and responses to the
 * given logger.
 *
 * Requests and responses are logged at debug level. If fetch throws, the
 * error is logged at warn level and the exception is propagated.
 *
 * @param logger Logger to send the logged messages to
 * @param fetch Fetch function to wrap
 * @returns A wrapper to fetch that logs requests and responses
 */
export const loggedFetch = <F extends FetchFn>(logger: FastifyBaseLogger, fetchFn: F): F => {
	const fn = async (...args: [RequestInfo | URL, RequestInit?]): Promise<Response> => {
		const input = args[0];
		const init = args[1];
		logger.debug({ input, init }, 'upstream request');
		const before = new Date();
		try {
			// make sure we pass all arguments, since the fetch functions from mesh
			// do use extra ones
			const resp = await fetchFn(...args);
			const text = await resp.text();
			const duration = new Date().getDate() - before.getDate();
			logger.debug(
				{ headers: Object.fromEntries(resp.headers.entries()), status: resp.status, text, duration: `${duration}ms` },
				'upstream response'
			);
			return new Response(text, resp);
		} catch (error: any) {
			logger.warn({ error }, 'upstream fetch failed');
			throw error;
		}
	};
	return fn as F;
};
