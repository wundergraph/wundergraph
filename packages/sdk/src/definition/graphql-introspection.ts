import { fetchFederationServiceSDL, isFederationService } from './federation-introspection';
import { configuration } from '../graphql/configuration';
import {
	buildClientSchema,
	buildSchema,
	getIntrospectionQuery,
	GraphQLSchema,
	parse,
	lexicographicSortSchema,
} from 'graphql';
import { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios';
import { ConfigurationVariableKind, DataSourceKind, HTTPHeader, HTTPMethod } from '@wundergraph/protobuf';
import { cleanupSchema } from '../graphql/schema';
import {
	applyNamespaceToDirectiveConfiguration,
	applyNameSpaceToFieldConfigurations,
	applyNameSpaceToGraphQLSchema,
	applyNameSpaceToTypeFields,
	generateTypeConfigurationsForNamespace,
} from './namespacing';
import { loadFile } from '../codegen/templates/typescript';
import { buildSubgraphSchema } from '@apollo/federation';
import * as https from 'https';
import { introspectWithCache } from './introspection-cache';
import { mapInputVariable, resolveVariable } from '../configure/variables';
import { buildMTLSConfiguration, buildUpstreamAuthentication, GraphQLApi, GraphQLIntrospection } from './index';
import { HeadersBuilder, mapHeaders } from './headers-builder';
import { Fetcher } from './introspection-fetcher';

export const resolveGraphqlIntrospectionHeaders = (headers?: { [key: string]: HTTPHeader }): Record<string, string> => {
	const baseHeaders: Record<string, string> = {
		'Content-Type': 'application/json',
		Accept: 'application/json',
	};
	if (headers !== undefined) {
		Object.keys(headers).forEach((key) => {
			if (headers[key].values.length === 1) {
				switch (headers[key].values[0].kind) {
					case ConfigurationVariableKind.STATIC_CONFIGURATION_VARIABLE:
						baseHeaders[key] = headers[key].values[0].staticVariableContent;
						break;
					case ConfigurationVariableKind.ENV_CONFIGURATION_VARIABLE:
						if (process.env[headers[key].values[0].environmentVariableName] !== undefined) {
							baseHeaders[key] = process.env[headers[key].values[0].environmentVariableName]!;
						} else if (headers[key].values[0].environmentVariableDefaultValue !== undefined) {
							baseHeaders[key] = headers[key].values[0].environmentVariableDefaultValue;
						}
						break;
				}
			}
		});
	}

	return baseHeaders;
};

export const introspectGraphql = async (introspection: GraphQLIntrospection): Promise<GraphQLApi> => {
	return introspectWithCache(introspection, async (introspection: GraphQLIntrospection): Promise<GraphQLApi> => {
		const headersBuilder = new HeadersBuilder();
		const introspectionHeadersBuilder = new HeadersBuilder();

		if (introspection.headers !== undefined) {
			introspection.headers(headersBuilder);
			introspection.headers(introspectionHeadersBuilder);
		}
		if (introspection.introspection?.headers !== undefined) {
			introspection.introspection?.headers(introspectionHeadersBuilder);
		}

		const headers = mapHeaders(headersBuilder);
		const introspectionHeaders = resolveGraphqlIntrospectionHeaders(mapHeaders(introspectionHeadersBuilder));

		let schema = await introspectGraphQLSchema(introspection, introspectionHeaders);
		schema = lexicographicSortSchema(schema);
		const federationEnabled = isFederationService(schema);
		const schemaSDL = cleanupSchema(
			schema,
			introspection.customFloatScalars || [],
			introspection.customIntScalars || []
		);
		const serviceSDL = !federationEnabled
			? undefined
			: await fetchFederationServiceSDL(resolveVariable(introspection.url), introspectionHeaders, {
					apiNamespace: introspection.apiNamespace,
			  });
		const serviceDocumentNode = serviceSDL !== undefined ? parse(serviceSDL) : undefined;
		const schemaDocumentNode = parse(schemaSDL);
		const graphQLSchema = buildSchema(schemaSDL);
		const { RootNodes, ChildNodes, Fields } = configuration(schemaDocumentNode, serviceDocumentNode);
		const subscriptionsEnabled = hasSubscriptions(schema);
		if (introspection.internal === true) {
			headers['X-WG-Internal-GraphQL-API'] = {
				values: [
					{
						kind: ConfigurationVariableKind.STATIC_CONFIGURATION_VARIABLE,
						staticVariableContent: 'true',
						environmentVariableName: '',
						environmentVariableDefaultValue: '',
						placeholderVariableName: '',
					},
				],
			};
		}
		return new GraphQLApi(
			applyNameSpaceToGraphQLSchema(schemaSDL, introspection.skipRenameRootFields || [], introspection.apiNamespace),
			[
				{
					Kind: DataSourceKind.GRAPHQL,
					RootNodes: applyNameSpaceToTypeFields(RootNodes, graphQLSchema, introspection.apiNamespace),
					ChildNodes: applyNameSpaceToTypeFields(ChildNodes, graphQLSchema, introspection.apiNamespace),
					Custom: {
						Fetch: {
							url: mapInputVariable(introspection.url),
							baseUrl: introspection.baseUrl ? mapInputVariable(introspection.baseUrl) : mapInputVariable(''),
							path: introspection.path ? mapInputVariable(introspection.path) : mapInputVariable(''),
							method: HTTPMethod.POST,
							body: mapInputVariable(''),
							header: headers,
							query: [],
							upstreamAuthentication: buildUpstreamAuthentication(introspection),
							mTLS: buildMTLSConfiguration(introspection),
							urlEncodeBody: false,
						},
						Subscription: {
							Enabled: subscriptionsEnabled,
							URL:
								introspection.subscriptionsURL !== undefined
									? mapInputVariable(introspection.subscriptionsURL)
									: typeof introspection.url === 'string'
									? mapInputVariable(introspection.url)
									: mapInputVariable(''),
							UseSSE: introspection.subscriptionsUseSSE !== undefined ? introspection.subscriptionsUseSSE : false,
						},
						Federation: {
							Enabled: federationEnabled,
							ServiceSDL: serviceSDL || '',
						},
						UpstreamSchema: schemaSDL,
					},
					Directives: applyNamespaceToDirectiveConfiguration(schema, introspection.apiNamespace),
				},
			],
			applyNameSpaceToFieldConfigurations(
				Fields,
				graphQLSchema,
				introspection.skipRenameRootFields || [],
				introspection.apiNamespace
			),
			generateTypeConfigurationsForNamespace(schemaSDL, introspection.apiNamespace),
			[]
		);
	});
};

const introspectGraphQLSchema = async (introspection: GraphQLIntrospection, headers?: Record<string, string>) => {
	if (introspection.loadSchemaFromString) {
		try {
			if (introspection.isFederation) {
				return buildSubgraphSchema(parse(loadFile(introspection.loadSchemaFromString)));
			}
			return buildSchema(loadFile(introspection.loadSchemaFromString));
		} catch (e: any) {
			throw new Error(
				`Loading schema from string failed for apiNamespace '${introspection.apiNamespace}'. Make sure the schema is valid and try again: ${e}`
			);
		}
	}
	try {
		return introspectGraphQLAPI(introspection, headers);
	} catch (e: any) {
		throw new Error(`Introspecting GraphQL API failed for apiNamespace '${introspection.apiNamespace}': ${e}`);
	}
};

const introspectGraphQLAPI = async (
	introspection: GraphQLIntrospection,
	headers?: Record<string, string>
): Promise<GraphQLSchema> => {
	const data = JSON.stringify({
		query: getIntrospectionQuery(),
		operationName: 'IntrospectionQuery',
	});

	let opts: AxiosRequestConfig = {
		headers: headers,
		'axios-retry': {
			onRetry: (retryCount: number, error: AxiosError, requestConfig: AxiosRequestConfig) => {
				let msg = `failed to run introspection query: method: ${requestConfig.method} url: ${requestConfig.url}`;
				if (introspection.apiNamespace) {
					msg += ` apiNamespace: ${introspection.apiNamespace}`;
				}
				msg += ` retryAttempt: ${retryCount}`;

				console.log(msg);
			},
		},
	};

	if (introspection.mTLS) {
		opts.httpsAgent = new https.Agent({
			key: resolveVariable(introspection.mTLS.key),
			cert: resolveVariable(introspection.mTLS.cert),
			rejectUnauthorized: !introspection.mTLS.insecureSkipVerify,
		});
	}

	let res: AxiosResponse | undefined;
	try {
		res = await Fetcher().post(resolveVariable(introspection.url), data, opts);
	} catch (e: any) {
		throw new Error(
			`introspection failed (url: ${introspection.url}, namespace: ${introspection.apiNamespace || ''}), error: ${
				e.message
			}`
		);
	}
	if (res === undefined) {
		throw new Error(
			`introspection failed (url: ${introspection.url}, namespace: ${introspection.apiNamespace || ''}), no response`
		);
	}
	if (res.status !== 200) {
		throw new Error(
			`introspection failed (url: ${introspection.url}, namespace: ${
				introspection.apiNamespace || ''
			}), response code: ${res.status}, message: ${res.statusText}`
		);
	}
	return buildClientSchema(res.data.data);
};

const hasSubscriptions = (schema: GraphQLSchema): boolean => {
	const subscriptionType = schema.getSubscriptionType();
	if (!subscriptionType) {
		return false;
	}
	const fields = subscriptionType.getFields();
	return Object.keys(fields).length !== 0;
};
