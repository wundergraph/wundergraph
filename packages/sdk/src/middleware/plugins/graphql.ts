import { FastifyPluginAsync } from 'fastify';
import fp from 'fastify-plugin';
import { GraphQLSchema } from 'graphql';
import {
	getGraphQLParameters,
	processRequest,
	renderGraphiQL,
	sendResult,
	shouldRenderGraphiQL,
	ExecutionContext as HelixExecutionContext,
} from 'graphql-helix';
import { BaseContext } from '../server';

export interface GraphQLServerConfig {
	serverName: string;
	schema: GraphQLSchema | Promise<GraphQLSchema>;
	apiNamespace?: string;
	enableGraphQLEndpoint?: boolean;
	routeUrl: string;
	// use baseContext to pass global data to the graphql context
	baseContext?: Object | Promise<Object>;
	// use the contextFactory to create a context for each request
	// the callback needs to be called with your object
	contextFactory?: (callback: (ctx: Object) => void) => Promise<void>;
	// by default, root fields will be prefixed with the api namespace
	// use skipRenameRootFields to skip this behaviour, e.g. if a field is already prefixed
	skipRenameRootFields?: string[];
	// implement resolverFactory to create a custom resolver
	customResolverFactory?: (executionContext: HelixExecutionContext & ExecutionContext) => Promise<{}>;
}

interface ExecutionContext {
	wundergraph: BaseContext;
}

const FastifyGraphQLPlugin: FastifyPluginAsync<GraphQLServerConfig> = async (fastify, config) => {
	const schema = await config.schema;
	const baseContext = await config.baseContext;

	fastify.route({
		method: ['GET', 'POST'],
		url: config.routeUrl,
		async handler(req, res) {
			const request = {
				body: req.body,
				headers: req.headers,
				method: req.method,
				query: req.query,
			};

			const pluginLogger = req.ctx.log.child({ server: config.serverName, plugin: 'graphql' });

			if (config.enableGraphQLEndpoint && shouldRenderGraphiQL(request)) {
				res.type('text/html');
				res.send(
					renderGraphiQL({
						endpoint: config.routeUrl,
					})
				);
			} else {
				const request = {
					body: req.body,
					headers: req.headers,
					method: req.method,
					query: req.query,
				};
				const { operationName, query, variables } = getGraphQLParameters(request);

				if (config.contextFactory) {
					await config.contextFactory(async (ctx) => {
						const result = await processRequest<ExecutionContext>({
							operationName,
							query,
							variables,
							request,
							schema,
							// @ts-ignore
							rootValueFactory: config.customResolverFactory,
							contextFactory: (): ExecutionContext => ({
								...baseContext,
								...ctx,
								wundergraph: {
									...req.ctx,
									log: pluginLogger,
								},
							}),
						});

						await sendResult(result, res.raw);
						res.sent = true;
					});
					return;
				}

				const result = await processRequest<ExecutionContext>({
					operationName,
					query,
					variables,
					request,
					schema,
					// @ts-ignore
					rootValueFactory: config.customResolverFactory,
					contextFactory: (): ExecutionContext => ({
						...baseContext,
						wundergraph: {
							...req.ctx,
							log: pluginLogger,
						},
					}),
				});

				await sendResult(result, res.raw);

				// skip async/await handling because we already sent the result with the raw http socket
				res.sent = true;
			}
		},
	});
};

export default fp(FastifyGraphQLPlugin, '3.x');
