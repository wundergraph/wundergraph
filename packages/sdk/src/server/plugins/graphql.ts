import { FastifyPluginAsync } from 'fastify';
import { GraphQLSchema } from 'graphql';
import {
	getGraphQLParameters,
	processRequest,
	renderGraphiQL,
	sendResult,
	shouldRenderGraphiQL,
	ExecutionContext as HelixExecutionContext,
} from 'graphql-helix';
import type { BaseRequestContext } from '../types';

export interface GraphQLServerConfig {
	serverName: string;
	schema: GraphQLSchema | Promise<GraphQLSchema>;
	apiNamespace?: string;
	enableGraphQLEndpoint?: boolean;
	routeUrl: string;
	// use baseContext to pass global data to the graphql context
	baseContext?: Object | Promise<Object>;
	// use the contextFactory to create a context for each request
	// the callback needs to be called with your object
	contextFactory?: (callback: (ctx: Object) => void) => Promise<void>;
	// by default, root fields will be prefixed with the api namespace
	// use skipRenameRootFields to skip this behaviour, e.g. if a field is already prefixed
	skipRenameRootFields?: string[];
	// implement resolverFactory to create a custom resolver
	customResolverFactory?: (executionContext: HelixExecutionContext & ExecutionContext) => Promise<{}>;
}

interface ExecutionContext {
	wundergraph: BaseRequestContext;
}

const FastifyGraphQLPlugin: FastifyPluginAsync<GraphQLServerConfig> = async (fastify, config) => {
	const schema = await config.schema;
	const baseContext = await config.baseContext;

	fastify.route({
		method: ['GET', 'POST'],
		url: config.routeUrl,
		async handler(req, reply) {
			const request = {
				body: req.body,
				headers: req.headers,
				method: req.method,
				query: req.query,
			};

			const pluginLogger = req.ctx.log.withFields({ server: config.serverName, plugin: 'graphql' });

			if (config.enableGraphQLEndpoint && shouldRenderGraphiQL(request)) {
				reply.type('text/html');
				reply.send(
					renderGraphiQL({
						endpoint: config.routeUrl,
					})
				);
			} else {
				// https://www.fastify.io/docs/latest/Reference/Reply/#hijack
				// We hand over the response handling to "graphql-helix"
				// No fastify hooks are called for the response.
				reply.hijack();

				const request = {
					body: req.body,
					headers: req.headers,
					method: req.method,
					query: req.query,
				};
				const { operationName, query, variables } = getGraphQLParameters(request);

				if (config.contextFactory) {
					await config.contextFactory(async (ctx) => {
						const result = await processRequest<ExecutionContext>({
							operationName,
							query,
							variables,
							request,
							schema,
							// @ts-ignore
							rootValueFactory: config.customResolverFactory,
							contextFactory: (): ExecutionContext => ({
								...baseContext,
								...ctx,
								wundergraph: {
									...req.ctx,
									log: pluginLogger,
								},
							}),
						});

						await sendResult(result, reply.raw);
					});
					return;
				}

				const result = await processRequest<ExecutionContext>({
					operationName,
					query,
					variables,
					request,
					schema,
					// @ts-ignore
					rootValueFactory: config.customResolverFactory,
					contextFactory: (): ExecutionContext => ({
						...baseContext,
						wundergraph: {
							...req.ctx,
							log: pluginLogger,
						},
					}),
				});

				await sendResult(result, reply.raw);
			}
		},
	});
};

export default FastifyGraphQLPlugin;
