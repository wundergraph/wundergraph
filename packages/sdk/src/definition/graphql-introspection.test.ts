import { introspect } from './index';
import fetch from 'node-fetch';
import nock from 'nock';
import axios from 'axios';
import { promises as fs } from 'fs';
import path from 'path';
import { assert } from 'chai';
import { buildSchema, lexicographicSortSchema, parse, printSchema } from 'graphql';
import { buildSubgraphSchema } from './graphql-introspection';

axios.defaults.adapter = require('axios/lib/adapters/http');

test('introspection via http', async () => {
	/*
	Uncomment this section if you want to update the test data from the real introspection endpoints.
	
	const fetchSDL = async (url: string, body: string) => {
	const introspection = await fetch(url, {
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
			},
			body,
		});
		return await introspection.json();
	};

	const sdlQuery = `{"query":"{_service{sdl}}"}`;
	const introspectionQuery = JSON.stringify({
		query: getIntrospectionQuery(),
		operationName: 'IntrospectionQuery',
	});

	const accountsSDL = await fetchSDL('http://localhost:4001/graphql', sdlQuery);
	const reviewsSDL = await fetchSDL('http://localhost:4002/graphql', sdlQuery);
	const productsSDL = await fetchSDL('http://localhost:4003/graphql', sdlQuery);
	const inventorySDL = await fetchSDL('http://localhost:4004/graphql', sdlQuery);
	const accountsIntrospection = await fetchSDL('http://localhost:4001/graphql', introspectionQuery);
	const reviewsIntrospection = await fetchSDL('http://localhost:4002/graphql', introspectionQuery);
	const productsIntrospection = await fetchSDL('http://localhost:4003/graphql', introspectionQuery);
	const inventoryIntrospection = await fetchSDL('http://localhost:4004/graphql', introspectionQuery);
	
	await fs.writeFile(path.join(__dirname, 'testdata','introspection.json'), JSON.stringify({
		accountsSDL,
		productsSDL,
		reviewsSDL,
		inventorySDL,
		accountsIntrospection,
		productsIntrospection,
		reviewsIntrospection,
		inventoryIntrospection,
	}, null, 2), {encoding: 'utf8', flag: 'w'});*/

	const data = await fs.readFile(path.join(__dirname, 'testdata', 'introspection.json'), { encoding: 'utf8' });
	const {
		accountsSDL,
		productsSDL,
		reviewsSDL,
		inventorySDL,
		accountsIntrospection,
		productsIntrospection,
		reviewsIntrospection,
		inventoryIntrospection,
	} = JSON.parse(data);

	const isSDL = (body: any) => body.query.includes('_service{sdl}');
	const isIntrospection = (body: any) => body.query.includes('IntrospectionQuery');

	nock('http://accounts.service')
		.post('/graphql', isSDL)
		.reply(200, accountsSDL)
		.post('/graphql', isIntrospection)
		.reply(200, accountsIntrospection)
		.persist();

	nock('http://products.service')
		.post('/graphql', isSDL)
		.reply(200, productsSDL)
		.post('/graphql', isIntrospection)
		.reply(200, productsIntrospection)
		.persist();

	nock('http://reviews.service')
		.post('/graphql', isSDL)
		.reply(200, reviewsSDL)
		.post('/graphql', isIntrospection)
		.reply(200, reviewsIntrospection)
		.persist();

	nock('http://inventory.service')
		.post('/graphql', isSDL)
		.reply(200, inventorySDL)
		.post('/graphql', isIntrospection)
		.reply(200, inventoryIntrospection)
		.persist();

	const generator = await introspect.federation({
		apiNamespace: 'federated',
		upstreams: [
			{
				url: 'http://accounts.service/graphql',
			},
			{
				url: 'http://reviews.service/graphql',
			},
			{
				url: 'http://products.service/graphql',
			},
			{
				url: 'http://inventory.service/graphql',
			},
		],
		introspection: {
			disableCache: true,
		},
	});

	const federatedApi = await generator({});
	expect(federatedApi).toMatchSnapshot();
});

test('introspection via string', async () => {
	const data = await fs.readFile(path.join(__dirname, 'testdata', 'introspection.json'), { encoding: 'utf8' });
	const { accountsSDL, productsSDL, reviewsSDL, inventorySDL } = JSON.parse(data);

	const generator = await introspect.federation({
		apiNamespace: 'federated',
		upstreams: [
			{
				url: 'http://accounts.service/graphql',
				loadSchemaFromString: accountsSDL.data._service.sdl,
			},
			{
				url: 'http://reviews.service/graphql',
				loadSchemaFromString: reviewsSDL.data._service.sdl,
			},
			{
				url: 'http://products.service/graphql',
				loadSchemaFromString: productsSDL.data._service.sdl,
			},
			{
				url: 'http://inventory.service/graphql',
				loadSchemaFromString: inventorySDL.data._service.sdl,
			},
		],
		introspection: {
			disableCache: true,
		},
	});

	const federatedApiFromString = await generator({});
	expect(federatedApiFromString).toMatchSnapshot();
});

test('build subgraph schema', async () => {
	const data = await fs.readFile(path.join(__dirname, 'testdata', 'introspection.json'), { encoding: 'utf8' });
	const { accountsSDL, productsSDL, reviewsSDL, inventorySDL } = JSON.parse(data);

	const accountsInput = accountsSDL.data._service.sdl;
	const productsInput = productsSDL.data._service.sdl;
	const reviewsInput = reviewsSDL.data._service.sdl;
	const inventoryInput = inventorySDL.data._service.sdl;

	const accountsSchema = printSchema(buildSubgraphSchema(parse(accountsInput)));
	const productsSchema = printSchema(buildSubgraphSchema(parse(productsInput)));
	const reviewsSchema = printSchema(buildSubgraphSchema(parse(reviewsInput)));
	const inventorySchema = printSchema(buildSubgraphSchema(parse(inventoryInput)));

	assert.equal(accountsSchema, prettifySchema(accountsSubgraphSchema));
	assert.equal(productsSchema, prettifySchema(productsSubgraphSchema));
	assert.equal(reviewsSchema, prettifySchema(reviewsSubgraphSchema));
	assert.equal(inventorySchema, prettifySchema(inventorySubgraphSchema));
});

const prettifySchema = (schema: string) => {
	return printSchema(lexicographicSortSchema(buildSchema(schema)));
};

const accountsSubgraphSchema = `directive @key(fields: String!) on OBJECT | INTERFACE

directive @extends on OBJECT | INTERFACE

directive @external on OBJECT | FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @provides(fields: String!) on FIELD_DEFINITION

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
  me: User
}

union _Entity = User

scalar _Any

type _Service {
  """
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  """
  sdl: String
}

type User {
  id: ID!
  name: String
  username: String
}`;

const productsSubgraphSchema = `directive @key(fields: String!) on OBJECT | INTERFACE

directive @extends on OBJECT | INTERFACE

directive @external on OBJECT | FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @provides(fields: String!) on FIELD_DEFINITION

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
  topProducts(first: Int = 5, random: Boolean): [Product]
}

union _Entity = Product

scalar _Any

type _Service {
  """
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  """
  sdl: String
}

type Product {
  upc: String!
  name: String
  price: Int
  weight: Int
}

type Subscription {
  updatedPrice: Product!
}

type Mutation {
  setPrice(upc: String!, price: Int!): Product
}`;

const reviewsSubgraphSchema = `directive @key(fields: String!) on OBJECT | INTERFACE

directive @extends on OBJECT | INTERFACE

directive @external on OBJECT | FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @provides(fields: String!) on FIELD_DEFINITION

type Review {
  id: ID!
  body: String
  author: User
  product: Product
}

type User {
  id: ID!
  username: String
  reviews: [Review]
}

type Product {
  upc: String!
  reviews: [Review]
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}

union _Entity = Review | User | Product

scalar _Any

type _Service {
  """
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  """
  sdl: String
}`;

const inventorySubgraphSchema = `directive @key(fields: String!) on OBJECT | INTERFACE

directive @extends on OBJECT | INTERFACE

directive @external on OBJECT | FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @provides(fields: String!) on FIELD_DEFINITION

type Product {
  upc: String!
  weight: Int
  price: Int
  inStock: Boolean
  shippingEstimate: Int
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}

union _Entity = Product

scalar _Any

type _Service {
  """
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  """
  sdl: String
}`;
