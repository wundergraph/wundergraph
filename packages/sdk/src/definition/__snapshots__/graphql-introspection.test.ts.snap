// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`introspection via http 1`] = `[Function]`;

exports[`introspection via string 1`] = `[Function]`;
